src/unix/async.c

```diff
@@ -62,32 +65,27 @@ int uv_async_send(uv_async_t* handle) {
     return 0;
 
   /* Tell the other thread we're busy with the handle. */
-  if (cmpxchgi(&handle->pending, 0, 1) != 0)
+  if (cmpxchgi(&handle->pending, 0, ASYNC_BUSY | ASYNC_PENDING) & ASYNC_PENDING)
     return 0;
 
   /* Wake up the other thread's event loop. */
   uv__async_send(handle->loop);
 
-  /* Tell the other thread we're done. */
-  if (cmpxchgi(&handle->pending, 1, 2) != 1)
-    abort();
+  /* Tell the other thread we're done by clearing BUSY FLAG */
+  cmpxchgi(&handle->pending, ASYNC_BUSY | ASYNC_PENDING, ASYNC_PENDING);
+  cmpxchgi(&handle->pending, ASYNC_BUSY, 0);
```

Comment:
Calling cmpxchgi() without checking the return value doesn't look right to me.

Neither does calling it twice. The first call sets it to PENDING, the second to zero but _only if_ it's BUSY. How?

Reply:
I know this way is not elegant, this two cmpxchgi just does atomic bit clear of PENDING flag.

Can I define a 'busy' field in UV_ASYNC_PRIVATE_FIELDS in unix.h ?  if so, the atomic operation can be done with single cmpxchgi.

I'll modify the PR according to your suggestions

Reply:
I make a PR with extra busy field in UV_ASYNC_PRIVATE_FIELDS , please check #2654.
I think it's more elegant and revised according to your review comments here.